framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -381,9 +379,85 @@ public void processClassTree(ClassTree classTree) {
 
         checkExtendsImplements(classTree);
 
+        checkPolymorphicClass(classTree);
+
         super.visitClass(classTree, null);
     }
 
+    /**
+     * Throws an error if {@code classTree} has polymorphic fields but is not annotated with
+     * {@code @HasQualifierParameter}.
+     */
+    private void checkPolymorphicClass(ClassTree classTree) {
+        List<? extends Tree> members = classTree.getMembers();
+        Set<? extends AnnotationMirror> topAnnotations =
+                atypeFactory.getQualifierHierarchy().getTopAnnotations();
+
+        for (Tree mem : members) {
+            if (mem.getKind() == Tree.Kind.VARIABLE) {
+                AnnotatedTypeMirror fieldAnno = atypeFactory.getAnnotatedType(mem);
+                for (AnnotationMirror top : topAnnotations) {
+                    AnnotationMirror polyAnnoInHierarchy =
+                            atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+                    if (polyAnnoInHierarchy != null) {
+                        PolyTypeScanner polyScanner = new PolyTypeScanner();
+                        if (polyScanner.visit(fieldAnno)) {
+                            Element classTreeElement = TreeUtils.elementFromTree(classTree);
+                            List<? extends AnnotationMirror> classAnnotations =
+                                    classTreeElement.getAnnotationMirrors();
+                            if (!AnnotationUtils.containsSameByName(
```

Comment:
Please use `hasQualifierParameter` methods in AnnotatedTypeFactory instead of directly looking for `HasQualifierParameter`.